presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java

```diff
@@ -180,11 +180,11 @@ public NodeSelector createNodeSelector(ConnectorId connectorId)
         }
     }
 
-    public static List<Node> selectNodes(int limit, Iterator<Node> candidates)
+    public static List<Node> selectNodes(int limit, ResettableRandomizedIterator<Node> candidates)
```

Comment:
Wow . I really feel we just want to use a `List<Node>` (and do `ImmutableList.copyOf` if the original input is a `Set` .

Reply:
Apparently we were trying to optimize randomization. So instead of shuffling all nodes everytime, only the number of nodes needed can be shuffled.

Reply:
By using `ResettableRandomizedIterator`, it already defeat the purpose since it will do a copy into `ArrayList` in constructor :)

Reply:
> By using `ResettableRandomizedIterator`, it already defeat the purpose since it will do a

We do a copy, but we don't call to the random number generator unless needed. Not sure how big a problem is that. But since it was optimized that way - let's keep it as is.